File: 4일차/길웅/타겟넘버.py
# 프로그래머스 '타겟 넘버'
# https://programmers.co.kr/learn/courses/30/lessons/43165

def dfs(numbers:list, target:int, idx:int, number:int):
    if idx >= len(numbers):
        if number == target:
            return 1
        return 0
    
    return dfs(numbers, target, idx + 1, number + numbers[idx]) + dfs(numbers, target, idx + 1, number - numbers[idx])


def solution(numbers, target):
    answer = dfs(numbers, target, 0, 0)
    return answer

File: 4일차/현동/타겟넘버.js
// https://programmers.co.kr/learn/courses/30/lessons/43165

function solution(numbers, target) {
  let answer = 0;

  function dfs(total, level) {
    if (level == numbers.length) {
      if (total == target) {
        ++answer;
      }
      return;
    }

    dfs(total + numbers[level], level + 1);
    dfs(total - numbers[level], level + 1);
  }

  dfs(numbers[0], 1);
  dfs(-numbers[0], 1);

  return answer;
}
